src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -132,16 +133,132 @@ fun inferExpressionType(expr: RsExpr): Ty {
 
         is RsArrayExpr -> inferArrayType(expr)
 
+        is RsRangeExpr -> {
+            val el = expr.exprList;
+            val dot2 = expr.dotdot;
+            val dot3 = expr.dotdotdot;
+
+            val (rangeName, rangeTy) = when {
+                dot2 != null && el.size == 0 -> "RangeFull" to null
+                dot2 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot2.startOffsetInParent) {
+                        "RangeFrom" to e.type
+                    } else {
+                        "RangeTo" to e.type
+                    }
+                }
+                dot2 != null && el.size == 2 -> {
+                    "Range" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+                dot3 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot3.startOffsetInParent) {
+                        return TyUnknown
+                    } else {
+                        "RangeToInclusive" to e.type
+                    }
+                }
+                dot3 != null && el.size == 2 -> {
+                    "RangeInclusive" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+
+                else -> return TyUnknown //Should never happen
+            }
+
+            val ty = tyFromAbsolutePath("std", "core", "ops::" + rangeName, expr);
+
+            if (rangeTy != null) {
+                val typeParameter = ty.getTypeParameter("Idx") ?: return ty
+                ty.substitute(mapOf(typeParameter to rangeTy))
+            } else {
+                ty
+            }
+        }
+
         is RsIndexExpr -> {
             val containerType = expr.containerExpr?.type ?: return TyUnknown
             val indexType = expr.indexExpr?.type ?: return TyUnknown
             findIndexOutputType(expr.project, containerType, indexType)
         }
 
+        is RsMacroExpr -> {
+            if (expr.vecMacro != null) {
+                val elements = expr.vecMacro!!.vecMacroArgs?.exprList ?: emptyList()
+                var elementType: Ty = TyUnknown;
+                for (e in elements) {
+                    elementType = getMoreCompleteType(e.type, elementType);
+                }
+
+                val ty = tyFromAbsolutePath("std", "collections", "vec::Vec", expr);
```

Comment:
We have `langItems` file for various "well-known" types (not only literally lang-items). Let's move these there as well? 

I think that `tyFromAbsolutePath` could be a `private` implementation detail, and the API would be something like `findStdVec(m: Module)`.